packages/stex-react-renderer/src/lib/PerSectionQuiz.tsx

```diff
@@ -15,6 +22,87 @@ export function handleViewSource(problemUri: string) {
     if (sourceLink) window.open(sourceLink, '_blank');
   });
 }
+function getSections(tocElems: FTML.TOCElem[]): string[] {
+  const sectionIds: string[] = [];
+  for (const tocElem of tocElems) {
+    if (tocElem.type === 'Section') {
+      sectionIds.push(tocElem.uri);
+    }
+    if ('children' in tocElem) {
+      sectionIds.push(...getSections(tocElem.children));
+    }
+  }
+  return sectionIds;
+}
+export async function getAllConceptUrisForCourse(
+  courseToc: FTML.TOCElem[] | undefined
+): Promise<Set<string>> {
+  if (!courseToc) {
+    return new Set<string>();
+  }
+  const sectionInfos = getSections(courseToc);
+  const flattenSections = (
+    sections: SectionInfo | SectionInfo[] | undefined
+  ): { uri: string }[] => {
+    if (!sections) return [];
+    if (Array.isArray(sections)) {
+      return sections.flatMap((s) => [{ uri: s.uri }, ...flattenSections(s.children)]);
+    }
+    return [{ uri: sections.uri }, ...flattenSections(sections.children)];
+  };
+  const allSections = sectionInfos.map((uri) => ({ uri }));
+  const sectionUris = allSections.map((info) => info.uri);
+  const conceptUris = new Set<string>();
+
+  for (const sectionUri of sectionUris) {
+    const defidenda = await getDefiniedaInSection(sectionUri);
+    defidenda.forEach((d) => conceptUris.add(d.conceptUri));
+    const deps = await getSectionDependencies(sectionUri);
+    deps.forEach((uri) => conceptUris.add(uri));
+  }
+  return conceptUris;
+}
+
+async function getLoRelationConceptUris(problemUri: string): Promise<string[]> {
+  const query = getSparqlQueryForLoRelationToDimAndConceptPair(problemUri);
+  const result = await getQueryResults(query ?? '');
+
+  const conceptUris: string[] = [];
+
+  result?.results?.bindings.forEach((binding) => {
+    const raw = binding.relatedData?.value;
+    if (!raw) return;
+    const parts = raw.split('; ').map((p) => p.trim());
+    const poSymbolUris = parts
+      .filter((data) => data.startsWith('http://mathhub.info/ulo#po-symbol='))
+      .map((data) => decodeURIComponent(data.split('#po-symbol=')[1]));
+
+    conceptUris.push(...poSymbolUris);
+  });
+  return conceptUris;
+}
+
+export async function getSyllabusAndAdventurousProblems(
+  sectionUri: string,
+  courseToc?: FTML.TOCElem
+) {
+  const conceptUrisFromCourse = await getAllConceptUrisForCourse(
```

Comment:
All of this needs to move to backend. I would suggest updating the get-problems-by-section API to return a more complex object. It should return an array of objects, where each object represents a problem. And along with the problemId it should contain the category and labels associated with the problem